lib/cli-engine/config-array/ignore-pattern.js

```diff
@@ -0,0 +1,230 @@
+/**
+ * @fileoverview `IgnorePattern` class.
+ *
+ * `IgnorePattern` class has the set of glob patterns and the base path.
+ *
+ * It provides two static methods.
+ *
+ * - `IgnorePattern.createDefaultIgnore(cwd)`
+ *      Create the default predicate function.
+ * - `IgnorePattern.createIgnore(ignorePatterns)`
+ *      Create the predicate function from multiple `IgnorePattern` objects.
+ *
+ * It provides two properties and a method.
+ *
+ * - `patterns`
+ *      The glob patterns that ignore to lint.
+ * - `basePath`
+ *      The base path of the glob patterns. If absolute paths existed in the
+ *      glob patterns, those are handled as relative paths to the base path.
+ * - `getPatternsRelativeTo(basePath)`
+ *      Get `patterns` as modified for a given base path. It modifies the
+ *      absolute paths in the patterns as prepending the difference of two base
+ *      paths.
+ *
+ * `ConfigArrayFactory` creates `IgnorePattern` objects when it processes
+ * `ignorePatterns` properties.
+ *
+ * @author Toru Nagashima <https://github.com/mysticatea>
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const assert = require("assert");
+const path = require("path");
+const ignore = require("ignore");
+const debug = require("debug")("eslint:ignore-pattern");
+
+/** @typedef {ReturnType<import("ignore").default>} Ignore */
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Get the path to the common ancestor directory of given paths.
+ * @param {string[]} sourcePaths The paths to calculate the common ancestor.
+ * @returns {string} The path to the common ancestor directory.
+ */
+function getCommonAncestorPath(sourcePaths) {
```

Comment:
Non-blocker for this PR but wanted to float the idea: I wonder if it would make sense to create a new `lib/shared/path-utils.js` file for these kinds of helper methods? That way if other modules are doing some similar things (I'm thinking about `lib/shared/runtime-info.js`, though this might not be the exact same check there) or if we add new behavior in the future we can easily re-use some of this functionality.

Reply:
I want to keep the `shared` directory minimal. I.e., the directory includes only stuff that is shared with multiple public APIs. In my mind, each directory in `lib` is like different packages.

Reply:
Where do you think shared utilities should go?

Reply:
If it's actually shared, the `shared` directory.

Reply:
If "shared or not" is a significant problem, I propose renaming that directory to `lib/common`. I disagree that utilities should only be extracted into their own packages only if multiple *public* APIs depend on them. We should always aim for a DRY (Don't Repeat Yourself) codebase.

Reply:
Well, the purpose of https://github.com/eslint/eslint/pull/11555 was to reduce the files in `util` directory because it's hard to track who is using the files. If the file is used for only one *purpose*, the file should place at the directory for *the purpose*. This is similar to we love local variables rather than global variables. I really don't want to make a trash box of utilities again.

Reply:
That's the mind I put into the name `lib/shared`. If we use a thing for multiple purposes, yeah, we move the thing to `lib/shared` (or an independent package) gladly.

Reply:
My intention wasn’t to derail this PR - let’s table this discussion for now.